pkg/upstream/upstream.go

```diff
@@ -377,3 +378,54 @@ func (up *Upstream) shouldClose() bool {
 
 	return false
 }
+
+// Verify checks if the username and password are valid.
+func (up *Upstream) Verify(ctx context.Context, username, password string) error {
+	tidbs, err := fetchTiDBTopology(ctx, up.etcdCli.Unwrap())
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if len(tidbs) == 0 {
+		return errors.New("tidb instance not found in topology, please check if the tidb is running")
+	}
+
+	for _, tidb := range tidbs {
+		// connect tidb
+		host := fmt.Sprintf("%s:%d", tidb.IP, tidb.Port)
+		dsnStr := fmt.Sprintf("%s:%s@tcp(%s)/", username, password, host)
```

Comment:
I wonder it is safe or not to directly put user/pwd in the connection string?

Reply:
It depends on the security of the mysql connection. Ref [mysql doc](https://dev.mysql.com/doc/refman/8.0/en/encrypted-connections.html#:~:text=By%20default%2C%20MySQL%20programs%20attempt%20to%20connect%20using%20encryption%20if%20the%20server%20supports%20encrypted%20connections%2C).